specs: Ecotone / 4844 / Dencun

protolambda opened the pull request:
**Description**

This PR introduces EIP-4844 and L2-Dencun changes as draft for the Ecotone upgrade.

There are some placeholder specifications still, which I'd like to collaborate on with everyone involved in the design / implementation of these features:
- [x] The blob-encoding (@roberto-bayardo)
- [x] `L1Block` contract-upgrading (@danyalprout)
- [x] `L1Block` contract changes (@danyalprout @anikaraghu)
- [x] L1-data-fees specifications (@roberto-bayardo @K-Ho )

protolambda committed:
specs: Eclipse / 4844 / Dencun draft

ajsutton commented on specs/derivation.md:
```diff
@@ -940,6 +966,114 @@ entries.
+The `L1Block` is called in the very first transaction of the block,
+and parsed to retrieve the L1 block attributes.
+
+To not modify or interrupt the above system transaction behavior,
+the `L1Block` contract upgrade is executed in the last block *before* the scheduled update.
```
If this is done as a deposit transaction it will be included before any L2 tx in that block and they may be unable to access correct L1 price data. If the contract upgrade is backwards compatible and can return the current pre-blob L1 data after the upgrade that's fine - otherwise the upgrade may cause issues for L2 transactions included in the last block before the tx.

And more generally, it feels very weird to activate an upgrade the block before it activates. Sadly the L1Info tx really has to be the first in the block because a heap of code in various places has baked in that dependency (shame we didn't think of upgrade tx earlier but such is life).
If the L1Info deposit tx is unchanged I'd do the contract upgrade in the activation block (after the L1Info tx). If the L1Info tx is different, I think I'd use the old deposit tx in the activation block, then the upgrade tx and the new deposit tx format is used from activation block + 1. That's ugly but ensures no changes happen until the activation block and it's inline with how the dao fork was implemented across multiple blocks so it's not unprecedented.


I

trianglesphere replied:
I like the idea of an upgrade transaction, but in this case I don't love that we break the L1 Info contract for a block. It'd be possible but a lot of work to allow upgrade transactions before the L1 Info tx (ok to do b/c it's a HF).

tynes replied:
Note that this part of the spec is modified in https://github.com/ethereum-optimism/optimism/pull/8702 and it is out of date here until that PR is merged

ajsutton commented on specs/superchain-upgrades.md:
```diff
@@ -310,7 +311,32 @@ The Delta upgrade uses a *L2 block-timestamp* activation-rule, and is specified
+  - [EIP-5656: MCOPY - Memory copying instruction](https://eips.ethereum.org/EIPS/eip-5656)
+  - [EIP-6780: SELFDESTRUCT only in same transaction](https://eips.ethereum.org/EIPS/eip-6780)
+  - [EIP-7516: BLOBBASEFEE opcode](https://eips.ethereum.org/EIPS/eip-7516)
+- Deneb (Consensus Layer): *not applicable to L1*
```
?
```suggestion
- Deneb (Consensus Layer): *not applicable to L2*
```

ajsutton reviewed (commented):
Looks good generally - I'm a bit concerned about upgrading the L1 info contract a block early but there aren't any great transactions so I could be convinced.

Given the L1 attributes predeployed contract is changing, we probably should update the [L1 Attributes Deposited Transaction](https://github.com/ethereum-optimism/optimism/blob/9ff3ebb3983be52c3ca189423ae7b4aec94e0fde/specs/deposits.md#l1-attributes-deposited-transaction) section to note the different call data that's required.  The definition is actually very generic but currently points to the pre-Eclipse contract and could easily be misleading if we don't at least note that the contract call changes as part of Eclipse activation.

protolambda committed:
specs: fix L1/L2 typo in specs/superchain-upgrades.md

Co-authored-by: Adrian Sutton <adrian@oplabs.co>

protolambda committed:
specs: 4844 blob encoding

protolambda committed:
specs: fix typo in Eclipse derivation spec

Co-authored-by: Evan Richard <5766842+EvanJRichard@users.noreply.github.com>

Commit:
Spec updates for Eclipse upgrade:

- l1 block attribute changes

- l1 cost func changes

- GasPriceOracle changes

- SystemConfig contract changes

- SystemConfig log event processing changes

Commit:
clarify blob calldata is to be ignored during chain derivation

danyalprout committed:
Update ecotone contract deployment

danyalprout committed:
combine cast commands / specify no-tx-pool

danyalprout committed:
Update from address on proxy calls

danyalprout committed:
Fix linting

danyalprout committed:
Add setEcotone transaction

Commit:
Ecotone spec updates for system_config

EvanJRichard commented on specs/derivation.md:
```diff
@@ -479,6 +483,27 @@ for each transaction:
 Each data-transaction is versioned and contains a series of [channel frames][g-channel-frame] to be read by the
 Frame Queue, see [Batch Submission Wire Format][wire-format].
 
+#### Eclipse: Blob Retrieval
+
+With the Eclipse upgrade the retrieval stage is extended to support an additional DA source: [EIP-4844] blobs.
+Upon each traversed transaction in the batch inbox:
```
```suggestion
#### Ecotone: Blob Retrieval

With the Ecotone upgrade the retrieval stage is extended to support an additional DA source: [EIP-4844] blobs.
Upon each traversed transaction in the batch inbox:
```

EvanJRichard commented on specs/derivation.md:
```diff
@@ -940,6 +966,114 @@ entries.
+- `data`: bytecode TBD, a multi-call:
+  - deploy new contract implementation
+  - upgrade of `L1Block` through `ProxyAdmin`
+- `sourceHash`: `0x7dc74874297a8937186fdbec57ad344647a522de456088557e5fdeda88f66ddd`,
+  computed with the "Upgrade-deposited" type, with `intent = "Eclipse: L1Block upgrade"`
```
```suggestion
- `sourceHash`: `0x7795a90486cc207315616d57bdbe5ca0ad63b22b5ba7fe087d11774f5de6e10b`,
  computed with the "Upgrade-deposited" type, with `intent = "Ecotone: L1Block upgrade"`
```

EvanJRichard commented on specs/derivation.md:
````diff
@@ -940,6 +966,114 @@ entries.
+
+Verify `sourceHash`:
+
+```bash
+# compute intent hash:
+cast keccak "Eclipse: L1Block upgrade"
+# 0x831b745c7397f93704ae55eb0100bf3c56fe9e304d3f21c1a93ec25f736fea26
+
+# source hash type:
+# 0x0000000000000000000000000000000000000000000000000000000000000002
+
+# compute source hash:
+cast keccak 0x0000000000000000000000000000000000000000000000000000000000000002831b745c7397f93704ae55eb0100bf3c56fe9e304d3f21c1a93ec25f736fea26
+# 0x7dc74874297a8937186fdbec57ad344647a522de456088557e5fdeda88f66ddd
+```
````
```suggestion
```bash
# compute intent hash:
cast keccak "Ecotone: L1Block upgrade"
# 0xaf2b20ee05be9fc3f0712050591a5f8988f94b56cdf48842863a773b76634fde

# source hash type:
# 0x0000000000000000000000000000000000000000000000000000000000000002

# compute source hash:
cast keccak 0x0000000000000000000000000000000000000000000000000000000000000002af2b20ee05be9fc3f0712050591a5f8988f94b56cdf48842863a773b76634fde
# 0x7795a90486cc207315616d57bdbe5ca0ad63b22b5ba7fe087d11774f5de6e10b
```
```

EvanJRichard commented on specs/derivation.md:
````diff
@@ -940,6 +966,114 @@ entries.
+# 0x7dc74874297a8937186fdbec57ad344647a522de456088557e5fdeda88f66ddd
+```
+
+#### Eclipse: Beacon block roots contract deployment (EIP-4788)
````
```suggestion
#### Ecotone: Beacon block roots contract deployment (EIP-4788)
```

EvanJRichard commented on specs/derivation.md:
````diff
@@ -940,6 +966,114 @@ entries.
+
+Verify `sourceHash`:
+
+```bash
+# compute intent hash:
+cast keccak "Eclipse: beacon block roots contract deployment"
+# 0x4e73a20ffe4a8330eb1f726862f4b062301e73d081c6d3824a6e0bd6428697fe
+
+# source hash type:
+# 0x0000000000000000000000000000000000000000000000000000000000000002
+
+# compute source hash:
+cast keccak 0x00000000000000000000000000000000000000000000000000000000000000024e73a20ffe4a8330eb1f726862f4b062301e73d081c6d3824a6e0bd6428697fe
+# 0xfbcd78e2e9665570c3f73026d601053af3892bdd06292d7eaf3adf4a1ee1392f
+```
````
```suggestion
```bash
# compute intent hash:
cast keccak "Ecotone: beacon block roots contract deployment"
# 0xab0dfc96b47739a0ae1d415bbfaae79ebb1111861a3b7cfbbaa6ca4a9e618357

# source hash type:
# 0x0000000000000000000000000000000000000000000000000000000000000002

# compute source hash:
cast keccak 0x0000000000000000000000000000000000000000000000000000000000000002ab0dfc96b47739a0ae1d415bbfaae79ebb1111861a3b7cfbbaa6ca4a9e618357
# 0x69b763c48478b9dc2f65ada09b3d92133ec592ea715ec65ad6e7f3dc519dc00c
```
```

EvanJRichard commented on specs/derivation.md:
```diff
@@ -940,6 +966,114 @@ entries.
+- `data`:
+  `0x60618060095f395ff33373fffffffffffffffffffffffffffffffffffffffe14604d57602036146024575f5ffd5b5f35801560495762001fff810690815414603c575f5ffd5b62001fff01545f5260205ff35b5f5ffd5b62001fff42064281555f359062001fff015500`
+- `isSystemTx`: `false`, as per the Regolith upgrade, even the system-generated transactions spend gas.
+- `sourceHash`: `0xfbcd78e2e9665570c3f73026d601053af3892bdd06292d7eaf3adf4a1ee1392f`,
```
```suggestion
- `sourceHash`: `0x69b763c48478b9dc2f65ada09b3d92133ec592ea715ec65ad6e7f3dc519dc00c`,
```

danyalprout committed:
Reduce gas limits, specify more accurate transaction descriptions

tynes committed:
Merge pull request #8660 from roberto-bayardo/blob-calldata

clarify blob batcher transaction handling in chain derivation

tynes committed:
Merge pull request #8708 from roberto-bayardo/system-config

Ecotone spec updates for system_config

tynes committed:
Merge pull request #8702 from danyalprout/flesh-out-network-upgrades

Specify L1Block + GasPriceOracle network txns

tynes commented on specs/deposits.md:
```diff
@@ -313,6 +316,16 @@ After running `pnpm build` in the `packages/contracts-bedrock` directory, the by
 the genesis file will be located in the `deployedBytecode` field of the build artifacts file at
 `/packages/contracts-bedrock/forge-artifacts/L1Block.sol/L1Block.json`.
 
+#### Eclipse L1Block upgrade
```
This section is out of date

danyalprout replied:
I'll put in a PR to update this 🙏

tynes commented:
> Looks good generally - I'm a bit concerned about upgrading the L1 info contract a block early but there aren't any great transactions so I could be convinced.

We have changed the functionality here such that the hf block contains the contract upgrade and then the next block contains the new call

tynes committed:
Merge pull request #8737 from roberto-bayardo/l1-block-attributes

spec out Ecotone L1 block attributes, l1 cost func, and gas price oracle changes

protolambda committed:
specs: rename Eclipse to Ecotone

danyalprout committed:
Update the Ecotone L1Block upgrade spec

danyalprout committed:
Fix set ecotone source hash

protolambda committed:
specs: 4844 blob encoding v2

danyalprout committed:
Update specs/deposits.md

Co-authored-by: protolambda <proto@protolambda.com>

protolambda committed:
specs: clarify data reads

protolambda commented on specs/predeploys.md:
```diff
@@ -231,17 +231,26 @@ the L1 portion of the fee. This fee pays for using L1 as a data availability
-The values used to compute the L2 portion of the fee are:
+The values used to compute the L1 portion of the fee prior to the Ecotone upgrade are:
 
 - scalar
 - overhead
 - decimals
```
TODO: need to clarify that these revert after the ecotone upgrade.
The `decimals` value is hardcoded, but I think it should also revert, since we don't use it anymore.

protolambda replied:
tracked here now: https://github.com/ethereum-optimism/protocol-quest/issues/73

protolambda committed:
Merge pull request #8744 from danyalprout/update-l1-block-info

Update l1 block info

protolambda committed:
Merge pull request #8745 from ethereum-optimism/4844-encoding-v2

specs: 4844 blob encoding v2

protolambda committed:
Merge branch '4844-blob-encoding' into eclipse-specs

maurelian commented on specs/deposits.md:
```diff
@@ -241,19 +246,41 @@ This transaction MUST have the following values:
+[ABI][ABI] encoded call to the `setL1BlockValues()` function with correct values associated with
+the corresponding L1 block (cf.  [reference implementation][l1-attr-ref-implem]).
+
+If the Ecotone upgrade is active, then `data` is instead a call to the `setL1BlockValuesEcotone()`
```
Nit: this name feels a bit clunky. 

Changing the arguments will allow this to exist alongside the existing function, is the `Ecotone` postfix desirable?

roberto-bayardo replied:
The thinking here was to make it clear the functions are in fact independent and post-Ecotone upgrade you must call the  new one.

maurelian replied:
Sounds reasonable. I won't block on this, but I would be a little bit worried about having a lot of this upgrade name post-fix scar tissue kicking around though.
I'd be interested in hear thoughts from other reviewers as they drive by. :)

tynes replied:
The idea is that we postfix everything with the hardfork name starting with bedrock. This gives a very clear and forwards compatible way of doing things across the whole codebase. fwiw, the geth codebase follows a similar pattern

tynes commented on specs/derivation.md:
```diff
@@ -940,6 +1049,232 @@ entries.
+
+To not modify or interrupt the system behavior around gas computation, this block will not include any sequenced
+transactions by setting `noTxPool: true`.
+
```
We need to specify the first `setL1BlockValuesEcotone` call to hold the invariant that the view of the system config matches what is on L1. Per https://github.com/ethereum-optimism/optimism/pull/8723, the current scalar value should be migrated to the `baseFeeScalar` and the `blobBaseFeeScalar` should be 0

tynes commented on specs/superchain-upgrades.md:
```diff
@@ -308,10 +309,35 @@ The Delta upgrade consists of a single consensus-layer feature: [Span Batches](.
+    - [The Beacon roots contract deployment is automated](./derivation.md#ecotone-beacon-block-roots-contract-deployment-eip-4788)
+  - [EIP-5656: MCOPY - Memory copying instruction](https://eips.ethereum.org/EIPS/eip-5656)
+  - [EIP-6780: SELFDESTRUCT only in same transaction](https://eips.ethereum.org/EIPS/eip-6780)
+  - [EIP-7516: BLOBBASEFEE opcode](https://eips.ethereum.org/EIPS/eip-7516)
```
This is underspecified - when you say not applicable to L2, does that mean it will always return 0?

protolambda replied:
My intention was to leave it functional, but let it return what it would have returned if there was never any blob tx included. We should probably explain in the spec that it then actually return `1`, not `0`. When running `CalcBlobFee(excessBlobGas=0)` that's what you get in geth.

tynes replied:
https://github.com/ethereum-optimism/optimism/pull/8830

tynes commented on specs/system_config.md:
```diff
@@ -89,6 +103,10 @@ The contained log events are filtered and processed as follows:
   - type `1`: `overhead` and `scalar` overwrite, as two packed `uint256` entries.
   - type `2`: `gasLimit` overwrite, as `uint64` payload.
   - type `3`: `unsafeBlockSigner` overwrite, as `address` payload.
+  - type `4`: `l1BasefeeScalar` and `l1BlobBasefeeScalar` overwrite, as two packed `uint32`
```
Perhaps it would be good to clarify that the evm will still pad to 32 bytes so the values will be in the least significant bytes tightly packed together in 8 bytes. It could be good to clarify the byte offsets of the values and their endianness

tynes replied:
What is the rationale for using `encodePacked` when it still pads to 32 bytes? It makes the parsing logic a bit different for this handler compared to the others and future versions of `solc` want to remove `encodePacked`. It would still be possible to tightly pack them without that helper function if we did want to upgrade but just wondering since it doesn't save any data size

roberto-bayardo replied:
my understanding was if we used regular abi.Encode it would be 32 bytes per uint for 64 bytes total.

tynes replied:
ah you are correct
